lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -34,420 +25,546 @@
-            SmallVector<FieldRef, 4> fields;
-            forallRefersTo(
+            recordValueRefersToFieldRef(
                 sub.getInput(),
-                [&](FieldRef subBase) {
-                  isValid = true;
-                  fields.push_back(subBase.getSubField(index));
-                  return success();
-                },
-                false);
-            if (isValid) {
-              for (auto f : fields)
-                setValRefsTo(res, f);
-            }
+                sub.getInput().getType().get().getFieldID(sub.getIndex()),
+                sub.getResult());
           })
-          .Case<SubaccessOp>([&](SubaccessOp sub) {
-            FVectorType vecType = sub.getInput().getType();
-            auto res = sub.getResult();
-            bool isValid = false;
-            SmallVector<FieldRef, 4> fields;
-            forallRefersTo(
+          .Case<SubfieldOp>([&](SubfieldOp sub) {
+            recordValueRefersToFieldRef(
                 sub.getInput(),
-                [&](FieldRef subBase) {
-                  isValid = true;
-                  // The result of a subaccess can refer to multiple storage
-                  // locations corresponding to all the possible indices.
-                  for (size_t index = 0; index < vecType.getNumElements();
-                       ++index)
-                    fields.push_back(subBase.getSubField(
-                        1 + index * (vecType.getElementType().getMaxFieldID() +
-                                     1)));
-                  return success();
-                },
-                false);
-            if (isValid) {
-              for (auto f : fields)
-                setValRefsTo(res, f);
-            }
+                sub.getInput().getType().get().getFieldID(sub.getFieldIndex()),
+                sub.getResult());
           })
-          .Case<InstanceOp>(
-              [&](InstanceOp ins) { handleInstanceOp(ins, worklist); })
-          .Case<MemOp>([&](MemOp mem) {
-            if (!(mem.getReadLatency() == 0)) {
-              return;
-            }
-            for (auto memPort : mem.getResults()) {
-              if (!type_isa<FIRRTLBaseType>(memPort.getType()))
-                continue;
-              memPorts.insert(memPort);
-            }
+          .Case<SubaccessOp>([&](SubaccessOp sub) {
+            auto vecType = sub.getInput().getType().get();
+            auto input = sub.getInput();
+            auto result = sub.getResult();
+            // Flatten the subaccess. The result can refer to any of the
+            // elements.
+            for (size_t index = 0; index < vecType.getNumElements(); ++index)
+              recordValueRefersToFieldRef(input, vecType.getFieldID(index),
+                                          result);
           })
-          .Default([&](auto) {});
+          .Case<RefSubOp>([&](RefSubOp sub) {
+            size_t fieldID = TypeSwitch<FIRRTLBaseType, size_t>(
```

Comment:
Can we use hw::FieldIDTypeInterface?

Reply:
I'm not sure if we can, how to implement the `sub.getIndex()`. Am I missing something ?

Reply:
You are right. Can we use FIRRTLTypeSwith either way?